## Permutations, Combinations, Conditional Probability and Partitioning Complex Events



## What's the difference between a permutation and a combination?

## How many permutations are there for a standard padlock?
> Hint: (there are 40 numbers on a padlock and use 3 numbers.)

## How many unique 3 topping pizzas can you make from the following ingredients:
    * Mushrooms
    * Pepperoni
    * Onion
    * Peppers
    * Ham
    * Pineapple
    * Sausage
    * Olives
    
> Fun question: which is the worst?

## Explaining the Intuition Behind Combinations


How would you compute the number of 3-topping combinations from the ingredients above by hand? Explain the rational behind your computations.

## Conditional Probability

### When do we compute conditional probabilities? 

### Mushroom dataset

To discuss conditional probability, let's look at a modified version of the Mushroom dataset from UCI [here](https://archive.ics.uci.edu/ml/datasets/Mushroom). Each row in this dataset corresponds to one observation (one mushroom). 

The modified dataset includes 4 variables:

* **edible-poisonous**
    * This categorical variable can have one of two values: if the mushroom is edible, "edible". If not, "poisonous"

* **bruised**
    * This is a Boolean variable that can assume either one of two values, True or False.

* **gill-spacing**
    * This categorical variable can have one of three values: "close", "crowded", or "distant"
    
* **stalk-shape**
    * This categorical variable can have one of two values: "enlarging" or "tapering"
* **stalk-color-above-ring**
    * This categorical variable can have one of nine values:  "brown","buff","cinnamon","gray","orange", "pink","red","white" or "yellow"

* **stalk-color-below-ring**
    * This categorical variable can have one of nine values:  "brown","buff","cinnamon","gray","orange", "pink","red","white" or "yellow"

* **gill-color**
    * This categorical variable can have one of twelve values: "black","brown","buff","chocolate","gray", "green","orange","pink","purple","red", "white" or "yellow" 



In [1]:
import pandas as pd

df = pd.read_csv('Mushrooms_cleaned.csv')
df.head()

3376
df.describe()

edible-poisonous gill-spacing stalk-shape stalk-color-above-ring  \
count              8124         8124        8124                   8124   
unique                2            2           2                      9   
top              edible        close    tapering                  white   
freq               4208         6812        4608                   4464   

       stalk-color-below-ring gill-color bruised  
count                    8124       8124    8124  
unique                      9         12       2  
top                     white       buff   False  
freq                     4384       1728    4748

#### If you picked a row from this dataset at random, what is the probability it corresponds to a bruised mushroom? 

#### What is the probability you pick a row corresponding to a mushroom that is bruised _AND_ edible? 

In [2]:
df[(df['bruised']== True) & (df['edible-poisonous'] == 'edible')]
    

edible-poisonous gill-spacing stalk-shape stalk-color-above-ring  \
1              edible        close   enlarging                  white   
2              edible        close   enlarging                  white   
5              edible        close   enlarging                  white   
6              edible        close   enlarging                  white   
7              edible        close   enlarging                  white   
9              edible        close   enlarging                  white   
10             edible        close   enlarging                  white   
11             edible        close   enlarging                  white   
12             edible        close   enlarging                  white   
20             edible        close   enlarging                  white   
22             edible        close   enlarging                  white   
23             edible        close   enlarging                  white   
24             edible        close   enlarging                  white   
26             edible        close   enlarging                  white   
27             edible        close   enlarging                  white   
29             edible      crowded    tapering                  white   
30             edible        close   enlarging                  white   
32             edible        close   enlarging                  white   
33             edible        close   enlarging                  white   
34             edible        close   enlarging                  white   
35             edible      crowded    tapering                  white   
38             edible      crowded    tapering                  white   
39             edible        close   enlarging                  white   
40             edible        close   enlarging                  white   
41             edible        close   enlarging                  white   
44             edible        close   enlarging                  white   
45             edible        close   enlarging                  white   
46             edible        close   enlarging                  white   
47             edible        close   enlarging                  white   
48             edible        close   enlarging                  white   
...               ...          ...         ...                    ...   
6928           edible        close   enlarging                  white   
6941           edible        close   enlarging                  white   
7022           edible        close   enlarging                  white   
7063           edible        close   enlarging                  white   
7073           edible        close   enlarging                  white   
7116           edible        close   enlarging                  white   
7181           edible        close   enlarging                  white   
7300           edible        close   enlarging                  white   
7353           edible        close   enlarging                  white   
7360           edible        close   enlarging                  white   
7370           edible        close   enlarging                  white   
7377           edible        close   enlarging                  white   
7415           edible        close   enlarging                  white   
7422           edible        close   enlarging                  white   
7443           edible        close   enlarging                  white   
7481           edible        close   enlarging                  white   
7548           edible        close   enlarging                  white   
7597           edible        close   enlarging                  white   
7602           edible        close   enlarging                  white   
7673           edible        close   enlarging                  white   
7704           edible        close   enlarging                  white   
7705           edible        close   enlarging                  white   
7721           edible        close   enlarging                 

#### What is the probability of picking an edible mushroom given it is bruised, $P(edible|bruised)$? 

#### What is the probability of picking a bruised mushroom given it is edible, $P(\text{bruised | edible})$? 

### Intuition behind conditional probability: 

How do you compute the probability that mushrooms are edible given they are bruised? 

When you ask the question "what is the probability that the mushrooms are edible and bruised?", the sample space originally contains all 8124 rows of mushrooms. 

<img src="images/Image_72_Cond4.png" width="300">

However, to compute the probability that the mushrooms are edible given they are bruised, you need to consider the reduced size of the sample space. 

In the image above, S is the universe of all mushrooms in the dataset, A is the set of mushrooms that are edible, and B is the set of mushrooms that are bruised.

* When you ask the question "what is the probability that the mushrooms are edible given the mushrooms are bruised?", you have effectively reduced the size of the sample space to include only those mushrooms that are bruised. 

* Given that mushrooms are bruised, the only way for the mushrooms to be edible is for these mushrooms to fall in the intersection of the set of mushrooms that are edible _and_ the set of mushrooms that are bruised , $P(edible \cap bruised)$.  

* To account for the smaller sample space, you divide the probability mushrooms are edible and bruised by the probability the mushrooms are bruised: $$\large P(edible|bruised) = \frac{P(edible \cap bruised)}{P(bruised)}$$




## Partitioning Complex Events

You're not really a mushroom expert, but you can see a bunch of orange spots all over the mushroom in your hand. Given the data at your disposal, what's the probability that the mushroom is edible?



In [7]:
p_orange = (df[(df['gill-color'] == 'orange') | (df['stalk-color-above-ring'] == 'orange') | (df['stalk-color-below-ring'] == 'orange')]).shape[0]/df.shape[0]

p_orange

0.023633677991137372

In [9]:
p_edible_and_orange = ((df[(df['gill-color'] == 'orange') | 
                           (df['stalk-color-above-ring'] == 'orange') | 
                           (df['stalk-color-below-ring'] == 'orange') & 
                           (df['edible-poisonous'] == 'edible')]).shape[0]/df.shape[0])

p_edible_and_orange

0.023633677991137372

## Summary


In this lesson, you reviewed 4 major foundational concepts for probability: permutations, combinations, conditional probability and partitioning complex events. Remember that your standard padlock should be more accurately called a permutation lock! Order matters for permutations, whereas only the members of the set are important for combinations. Conditional probability investigates the odds of an event occurring given other information. In these instances, the universal set of possibilities reflects the given information. In the mushroom example, the probability of a mushroom being edible given that it is bruised can be computed by dividing the probability that it a mushroom is both edible AND bruised, by the probability that it is bruised. Mathematically: $$\large P(edible|bruised) = \frac{P(edible \cap bruised)}{P(bruised)}$$. Finally, you investigated partitioning complex events. Often, complex events can be broken into constituent parts, and the total probability can be calculated by combining these smaller events.

## Additional Resources

## Challenge Problem

Let's take some time and review questions like those from the [dsc-law-of-total-probability-lab](https://github.com/learn-co-curriculum/dsc-law-of-total-probability-lab).  

According to the CDC, [14% of Americans currently smoke, 15.8% of males and 12.2% of females](https://www.cdc.gov/tobacco/data_statistics/fact_sheets/adult_data/cig_smoking/index.htm). According the the American Lung Association, [men who smoke are 23 times more likely to smoke then never-smokers, and women are 13 times as likely](https://www.lung.org/lung-health-and-diseases/lung-disease-lookup/lung-cancer/resource-library/lung-cancer-fact-sheet.html). The American Cancer Society estimates that [the lifetime risk of developing lung cancer is 6.85% for males and 5.95% for females](https://www.cancer.org/cancer/cancer-basics/lifetime-probability-of-developing-or-dying-from-cancer.html). Currently, the census estimates that [women are 50.8% of the population](https://www.census.gov/quickfacts/fact/table/US/PST045218). 

What is the risk of lung cancer for non-smokers? Non-smoker males? Non-smoker females?

> To learn more about lung-cancer risks for non-smokers, see https://www.cancer.org/latest-news/why-lung-cancer-strikes-nonsmokers.html.

## Additional Conditional Probability Practice

What's the probability that a mushroom is poisonous if it has close gill spacing and a tapering stalk?

$$\large P(edible|close \cap tapering) = \frac{P(edible \cap close \cap tapering)}{P(close \cap tapering)}$$